File: src/libsokoengine/game/hexoban_tessellation.cpp
#include "hexoban_tessellation.hpp"

#include "pusher_step.hpp"
#include "snapshot.hpp"

using namespace std;

namespace sokoengine {
namespace game {

const Directions &HexobanTessellation::legal_directions() const {
  static const Directions retv = {Direction::LEFT,       Direction::RIGHT,
                                  Direction::NORTH_EAST, Direction::NORTH_WEST,
                                  Direction::SOUTH_EAST, Direction::SOUTH_WEST};
  return retv;
}

position_t HexobanTessellation::neighbor_position(position_t position,
                                                  const Direction &direction,
                                                  board_size_t width,
                                                  board_size_t height) const {
  position_t row = Y(position, width), column = X(position, width);
  switch (direction) {
  case Direction::LEFT:
    column += -1;
    break;
  case Direction::RIGHT:
    column += 1;
    break;
  case Direction::NORTH_EAST:
    column += row % 2;
    row -= 1;
    break;
  case Direction::NORTH_WEST:
    column -= (row + 1) % 2;
    row -= 1;
    break;
  case Direction::SOUTH_EAST:
    column += row % 2;
    row += 1;
    break;
  case Direction::SOUTH_WEST:
    column -= (row + 1) % 2;
    row += 1;
    break;
  default:
    throw invalid_argument(
      "Unsupported Direction received in HexobanTessellation neighbor_position!");
  }

  if (ON_BOARD(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::MAX_POS + 1;
}

PusherStep HexobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
  case io::Snapshot::l:
    return PusherStep(Direction::LEFT, false);
  case io::Snapshot::L:
    return PusherStep(Direction::LEFT, true);

  case io::Snapshot::r:
    return PusherStep(Direction::RIGHT, false);
  case io::Snapshot::R:
    return PusherStep(Direction::RIGHT, true);

  case io::Snapshot::u:
    return PusherStep(Direction::NORTH_WEST, false);
  case io::Snapshot::U:
    return PusherStep(Direction::NORTH_WEST, true);

  case io::Snapshot::d:
    return PusherStep(Direction::SOUTH_EAST, false);
  case io::Snapshot::D:
    return PusherStep(Direction::SOUTH_EAST, true);

  case io::Snapshot::n:
    return PusherStep(Direction::NORTH_EAST, false);
  case io::Snapshot::N:
    return PusherStep(Direction::NORTH_EAST, true);

  case io::Snapshot::s:
    return PusherStep(Direction::SOUTH_WEST, false);
  case io::Snapshot::S:
    return PusherStep(Direction::SOUTH_WEST, true);

  default:
    throw invalid_argument("Illegal PusherStep character in HexobanTessellation!");
  }
}

char HexobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
  case Direction::LEFT:
    return rv.is_push_or_pull() ? io::Snapshot::L : io::Snapshot::l;
  case Direction::RIGHT:
    return rv.is_push_or_pull() ? io::Snapshot::R : io::Snapshot::r;
  case Direction::NORTH_WEST:
    return rv.is_push_or_pull() ? io::Snapshot::U : io::Snapshot::u;
  case Direction::SOUTH_EAST:
    return rv.is_push_or_pull() ? io::Snapshot::D : io::Snapshot::d;
  case Direction::NORTH_EAST:
    return rv.is_push_or_pull() ? io::Snapshot::N : io::Snapshot::n;
  case Direction::SOUTH_WEST:
    return rv.is_push_or_pull() ? io::Snapshot::S : io::Snapshot::s;
  default:
    throw invalid_argument("Illegal PusherStep direction in HexobanTessellation!");
  }
}

} // namespace game
} // namespace sokoengine

File: src/sokoenginepy/game/hexoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, Tuple

from ..io import Snapshot
from .base_tessellation import BaseTessellation
from .config import Config, Direction
from .coordinate_helpers import index_1d, index_column, index_row, is_on_board_2d
from .utilities import inverted


class HexobanTessellation(BaseTessellation):
    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, Tuple[Direction, bool]]] = {
        Snapshot.l: (Direction.LEFT, False),
        Snapshot.L: (Direction.LEFT, True),
        Snapshot.r: (Direction.RIGHT, False),
        Snapshot.R: (Direction.RIGHT, True),
        Snapshot.u: (Direction.NORTH_WEST, False),
        Snapshot.U: (Direction.NORTH_WEST, True),
        Snapshot.d: (Direction.SOUTH_EAST, False),
        Snapshot.D: (Direction.SOUTH_EAST, True),
        Snapshot.n: (Direction.NORTH_EAST, False),
        Snapshot.N: (Direction.NORTH_EAST, True),
        Snapshot.s: (Direction.SOUTH_WEST, False),
        Snapshot.S: (Direction.SOUTH_WEST, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[Tuple[Direction, bool], str]] = inverted(
        _CHR_TO_PUSHER_STEP
    )

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)

        if direction == Direction.LEFT:
            column -= 1
        elif direction == Direction.RIGHT:
            column += 1
        elif direction == Direction.NORTH_EAST:
            column += row % 2
            row -= 1
        elif direction == Direction.NORTH_WEST:
            column -= (row + 1) % 2
            row -= 1
        elif direction == Direction.SOUTH_EAST:
            column += row % 2
            row += 1
        elif direction == Direction.SOUTH_WEST:
            column -= (row + 1) % 2
            row += 1
        else:
            raise ValueError(
                f"Unsupported direction {direction} for {self.__class__.__name__}"
            )

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS
